### ISSUE TITLE
Auto Focus not working / No option to set it

### ISSUE THREAD

**Issue by dhartwich1991** [googlesamples/android-vision#2]
Just checked out the code for the multi-tracker application and compiled it to my device. When I tried to scan barcodes it failed, because there is no auto focus which makes it really hard to scan barcodes without any problems. The CameraSource class does not provide any options to set an auto focus option.

**Comment by revanmj** [comment]
Same here. Small sized codes cannot be recognized because they are all completely blurred while camera is close or really tiny when it's away and image is sharp.

**Comment by xGary** [comment]
Same. Faces and some QR codes work fine but barcodes need autofocus to work as they're a lot more intricate.

**Comment by pm0733464** [comment]
Sorry about that.  We are working on adding an auto focus option to CameraSource and will let you know when a fix is available.

**Comment by Aeefire** [comment]
meanwhile, you can use this fix https://gist.github.com/Gericop/7de0b9fdd7a444e53b5a

it works for me until the library is updated.

**Comment by qbalsdon** [comment]
@Aeefire How did you invoke the method in the fix? Where did you put that method? What is the value of the focusMode parameter?

**Comment by crebstock** [comment]
Unless I'm missing something, this would also be the method to turn torch mode on.

**Comment by Gericop** [comment]
Yes, that's the way to do it. Actually, this method I provided only does the focus thing so the Vision API could **actually work**, but you can change it so that it returns the Camera instance (*or null on failure*), then use that to modify the camera's parameters (focus, flash, etc.).

**Comment by truizlop** [comment]
This solution does not work for me. It doesn't enable auto focus. I made sure that it is called when the camera is displaying. Besides, sometimes I am able to read 1D codes, but the raw value doesn't match the barcode value, and it returns different values every time I try to read the same barcode.

**Comment by pm0733464** [comment]
Can you post the code that you wrote to try to enable auto focus?

Note that some devices don't support camera auto focus (hardware limitation).  See the comment that I just added here:  https://gist.github.com/Gericop/7de0b9fdd7a444e53b5a

**Comment by Gericop** [comment]
What device are you testing on? It is possible your device doesn't support continuous auto focus.
Where did you put your cameraFocus(...) call? I put it in CameraSourcePreview's startIfReady() like this:

```java
private void startIfReady() throws IOException {
	if (mStartRequested && mSurfaceAvailable) {
		mCameraSource.start(mSurfaceView.getHolder());
		// ###### HERE IS THE FOCUS CALL ######
		MultiTrackerActivity.cameraFocus(mCameraSource, Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE);
		// ####################################
		if (mOverlay != null) {
			Size size = mCameraSource.getPreviewSize();
			int min = Math.min(size.getWidth(), size.getHeight());
			int max = Math.max(size.getWidth(), size.getHeight());
			if (isPortraitMode()) {
				// Swap width and height sizes when in portrait, since it will be rotated by
				// 90 degrees
				mOverlay.setCameraInfo(min, max, mCameraSource.getCameraFacing());
			} else {
				mOverlay.setCameraInfo(max, min, mCameraSource.getCameraFacing());
			}
			mOverlay.clear();
		}
		mStartRequested = false;
	}
}
```
You might also want to try to invoke the method with a direct action, like in a button's onClickListener.

**Comment by truizlop** [comment]
That's exactly the code I am using. Probably it is because my phone doesn't support auto focus. 2D barcodes are read correctly, but 1D, when they are read, provide different values for different scans.

**Comment by pm0733464** [comment]
We have recently open sourced the CameraSource class:

https://github.com/googlesamples/android-vision/blob/master/visionSamples/barcode-reader/app/src/main/java/com/google/android/gms/samples/vision/barcodereader/ui/camera/CameraSource.java

This version has an auto-focus option.

**Comment by thanhnguyen9294** [comment]
How to call  auto-focus option of CameraSource , i'm trying to call it but getting fail...now in my case  ..onPictureTaken resulting in very poor quality  images.. can anyone help me..??

**Comment by pm0733464** [comment]
Can you post your code that creates the CameraSource and the failure messages that you are getting?

**Comment by pm0733464** [comment]
**Update:**
The CameraSource implementation provided by Google Play Services 8.4 includes an auto focus option.

**Comment by ChristianBecker** [comment]
@pm0733464 Do you have any plans to extend the official `CameraSource` implementation by the flash mode option that is already available in the open-source `CameraSource` version?

**Comment by pm0733464** [comment]
We could include the flash mode if there is sufficient demand.

**Comment by Gericop** [comment]
I think the whole open source `CameraSource` could (*should?*) replace the official one as it offers more options to the devs.

**Comment by ChristianBecker** [comment]
Totally. Are there any downsides if the official `CameraSource` would be extended by the flash mode options?

**Comment by ChristianBecker** [comment]
@pm0733464 : Any progress on extending the official `CameraSource` by the flash mode options?

**Comment by naveeng915213** [comment]
@pm0733464: Hi am using  vision api to scan barcode, as autofocus is working in samsung tab 3.. I donno how to figure it out whats the problems. but when I installed the apps present store, those are well scanning and autofocusing in my samsung tablet, I donno why google vision api is failing to do so.. please can you help me out to fix this issue.

**Comment by onuryurtturk** [comment]
hi, is there any solution? its a major problem for us. Only got focus by clicking to screen. Its not a good way...
